rules/memory-leak/memory-leak.js

```diff
@@ -0,0 +1,181 @@
+}
+
+function getObjectNameForNode(node) {
+  if (isMemberExpressionWithIdentifier(node.parent.object) || hasNestedMember(node.parent) || isThisIdentifier(node.parent.object)) {
+    return node.parent.object.property.name;
+  } else if (isDotOn(node.parent)) {
+    return 'this';
+  } else if (isCallExpression(node.parent.object) && calleeIsIdentifier(node.parent)) {
+    const argsList = getArgumentsString(node.parent.object.arguments);
+    return `${node.parent.object.callee.name}(${argsList})`;
+  } else if (isCallExpression(node.parent.object) && isMemberExpressionWithObjectIdentifier(node.parent.object.callee)) {
+    const argsList = getArgumentsString(node.parent.object.arguments);
+    return `${node.parent.object.callee.object.name}.${node.parent.object.callee.property.name}(${argsList})`;
+  } else if (notSure2(node.parent)) {
+    return node.parent.object.name;
+  }
+}
+
+function getArgumentsString(args) {
+  return args.map(stringifyNode).join(', ');
+}
+
+function stringifyNode(node) {
+  if (node.type === 'MemberExpression' && isThisIdentifier(node)) {
+    return `this.${node.property.name}`;
+  } else if (isMemberExpressionWithObjectIdentifier(node)) {
+    return `${node.object.name}.${node.property.name}`;
+  } else if (node.type === 'Literal') {
+    return `'${node.value}'`;
+  } else if (node.type === 'Identifier') {
+    return node.name;
+  } else {
+    return undefined;
+  }
+}
+
+
+///////////////////////////// NODE CONDITIONALS /////////////////////////////
+
+// someObject.on('eventName', ...)
+function isEventSubscription(node) {
+  return node.name === 'on' &&
+    node.type === 'Identifier' &&
+    node.parent.type === 'MemberExpression' &&
+    node.parent.parent.type === 'CallExpression' &&
+    node.parent.parent.arguments.length === 2 &&
+    node.parent.parent.arguments[0].type === 'Literal' &&
+    node.parent.parent.arguments[0].value !== '$destroy';
+}
+
+// someObject.attribute
+function isMemberExpressionWithIdentifier(node) {
+  return node.type === 'MemberExpression' &&
+    node.property &&
+    node.property.type === 'Identifier';
+}
+
+// this.on(...)
+function isDotOn(node) {
+  return isMemberExpressionWithIdentifier(node.object) &&
+    node.property.name === 'on';
+}
+
+// func(value) or someObject.func(value)
+function isCallExpression(node) {
+  return node.type === 'CallExpression' &&
+    node.callee &&
+    node.arguments;
+}
+
+// func(value)
+function calleeIsIdentifier(node) {
+  return node.object.callee.type === 'Identifier';
+}
+
+// someObject.func(value)
+function isMemberExpressionWithObjectIdentifier(node) {
+  return isMemberExpressionWithIdentifier(node) &&
+    node.object &&
+    node.object.type === 'Identifier';
+}
+
+// someObject.member.nestedMember
+function hasNestedMember(node) {
+  return node.object.object &&
+    node.object.object.type === 'Identifier' &&
+    node.object.property &&
+    node.object.property.type === 'Identifier';
+}
+
+// Not sure...
```

Comment:
@szydan I'd like some guidance as to the names and examples for these node conditional functions; my knowledge of JavaScript language semantics isn't great, so what I have here might not be accurate.

Reply:
@daradermody First big thanks for taking care about this repo ❤️ 
About the names I am not sure myself as it is a while since I wrote it
Seems 
notSure2 ->  isThisIdentifier
isThisIdentifier -> isThisIdentifierInInExpression 

The best would be to add the test cases from #2 and see for which use case the particular rule fires 
then we might get better idea about the naming

Reply:
Made function name and comment clarifications based on what conditions were followed for each test example.